Serialization stack overflow

The game crashed with a stack overflow error. The stack trace was the following, repeated many times:

	at org.terasology.persistence.typeHandling.FutureTypeHandler.serialize(FutureTypeHandler.java:32)
	at org.terasology.persistence.typeHandling.coreTypes.RuntimeDelegatingTypeHandler.serializeViaDelegate(RuntimeDelegatingTypeHandler.java:162)
	at org.terasology.persistence.typeHandling.coreTypes.RuntimeDelegatingTypeHandler.serializeNonNull(RuntimeDelegatingTypeHandler.java:118)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.ObjectFieldMapTypeHandler.serializeNonNull(ObjectFieldMapTypeHandler.java:71)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)

It ran for a minute or so before this happened, using much more CPU than usual (but this was probably just the ocean falling into caves). I was using JS + FlowingLiquids (and some almost certainly irrelevant changes that would never have even executed under the circumstances).

I haven't been able to replicate it in any other worlds, and I can't load the one where I originally saw it because chunk generation times out.

```
14:16:47.532 [main] INFO  o.t.p.i.ReadWriteStorageManager - Auto Saving - Creating game snapshot
14:16:47.539 [main] INFO  o.t.p.i.ReadWriteStorageManager - Auto Saving - Snapshot created: Writing phase starts
14:16:47.686 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.RuntimeException: Saving failed
	at org.terasology.persistence.internal.ReadWriteStorageManager.checkSaveTransactionAndClearUpIfItIsDone(ReadWriteStorageManager.java:175)
	at org.terasology.persistence.internal.ReadWriteStorageManager.update(ReadWriteStorageManager.java:393)
	at org.terasology.engine.modes.StateIngame.update(StateIngame.java:179)
	at org.terasology.engine.TerasologyEngine.tick(TerasologyEngine.java:464)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:427)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:403)
	at org.terasology.engine.Terasology.main(Terasology.java:163)
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make field private jdk.internal.ref.Cleaner jdk.internal.ref.Cleaner.next accessible: module java.base does not "opens jdk.internal.ref" to unnamed module @5ebd56e9
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:340)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:280)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:176)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:170)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.lambda$getResolvedFields$1(ObjectFieldMapTypeHandlerFactory.java:96)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.getResolvedFields(ObjectFieldMapTypeHandlerFactory.java:84)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.create(ObjectFieldMapTypeHandlerFactory.java:56)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:384)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:321)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.lambda$create$0(ObjectFieldMapTypeHandlerFactory.java:59)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.create(ObjectFieldMapTypeHandlerFactory.java:56)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:384)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:321)
	at org.terasology.persistence.typeHandling.coreTypes.RuntimeDelegatingTypeHandler.serializeNonNull(RuntimeDelegatingTypeHandler.java:79)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.ObjectFieldMapTypeHandler.serializeNonNull(ObjectFieldMapTypeHandler.java:71)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.RuntimeDelegatingTypeHandler.serializeNonNull(RuntimeDelegatingTypeHandler.java:131)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.ObjectFieldMapTypeHandler.serializeNonNull(ObjectFieldMapTypeHandler.java:71)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.Serializer.serializeValue(Serializer.java:80)
	at org.terasology.persistence.serializers.ComponentSerializer.serialize(ComponentSerializer.java:292)
	at org.terasology.persistence.serializers.EntitySerializer.serializeEntityDelta(EntitySerializer.java:337)
	at org.terasology.persistence.serializers.EntitySerializer.serialize(EntitySerializer.java:161)
	at org.terasology.persistence.internal.EntityStorer.store(EntityStorer.java:67)
	at org.terasology.persistence.internal.EntityStorer.store(EntityStorer.java:57)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.terasology.persistence.internal.CompressedChunkBuilder.<init>(CompressedChunkBuilder.java:52)
	at org.terasology.persistence.internal.SaveTransaction.prepareCompressedChunkBuilders(SaveTransaction.java:232)
	at org.terasology.persistence.internal.SaveTransaction.prepareChunksPlayersAndGlobalStore(SaveTransaction.java:211)
	at org.terasology.persistence.internal.SaveTransaction.run(SaveTransaction.java:149)
	at org.terasology.utilities.concurrency.TaskProcessor.run(TaskProcessor.java:50)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
14:16:47.687 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
14:16:48.099 [main] ERROR o.terasology.engine.TerasologyEngine - Clean game shutdown after an uncaught exception failed
java.lang.reflect.InaccessibleObjectException: Unable to make field private jdk.internal.ref.Cleaner jdk.internal.ref.Cleaner.next accessible: module java.base does not "opens jdk.internal.ref" to unnamed module @5ebd56e9
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:340)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:280)
	at java.base/java.lang.reflect.Field.checkCanSetAccessible(Field.java:176)
	at java.base/java.lang.reflect.Field.setAccessible(Field.java:170)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.lambda$getResolvedFields$1(ObjectFieldMapTypeHandlerFactory.java:96)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.getResolvedFields(ObjectFieldMapTypeHandlerFactory.java:84)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.create(ObjectFieldMapTypeHandlerFactory.java:56)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:384)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:321)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.lambda$create$0(ObjectFieldMapTypeHandlerFactory.java:59)
	at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:684)
	at org.terasology.persistence.typeHandling.coreTypes.factories.ObjectFieldMapTypeHandlerFactory.create(ObjectFieldMapTypeHandlerFactory.java:56)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:384)
	at org.terasology.persistence.typeHandling.TypeHandlerLibrary.getTypeHandler(TypeHandlerLibrary.java:321)
	at org.terasology.persistence.typeHandling.coreTypes.RuntimeDelegatingTypeHandler.serializeNonNull(RuntimeDelegatingTypeHandler.java:79)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.ObjectFieldMapTypeHandler.serializeNonNull(ObjectFieldMapTypeHandler.java:71)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.RuntimeDelegatingTypeHandler.serializeNonNull(RuntimeDelegatingTypeHandler.java:131)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.coreTypes.ObjectFieldMapTypeHandler.serializeNonNull(ObjectFieldMapTypeHandler.java:71)
	at org.terasology.persistence.typeHandling.TypeHandler.serialize(TypeHandler.java:49)
	at org.terasology.persistence.typeHandling.Serializer.serializeValue(Serializer.java:80)
	at org.terasology.persistence.serializers.ComponentSerializer.serialize(ComponentSerializer.java:292)
	at org.terasology.persistence.serializers.EntitySerializer.serializeEntityDelta(EntitySerializer.java:337)
	at org.terasology.persistence.serializers.EntitySerializer.serialize(EntitySerializer.java:161)
	at org.terasology.persistence.internal.EntityStorer.store(EntityStorer.java:67)
	at org.terasology.persistence.internal.EntityStorer.store(EntityStorer.java:57)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at org.terasology.persistence.internal.CompressedChunkBuilder.<init>(CompressedChunkBuilder.java:52)
	at org.terasology.persistence.internal.ReadWriteStorageManager.deactivateChunk(ReadWriteStorageManager.java:330)
	at org.terasology.world.chunks.localChunkProvider.LocalChunkProvider.unloadChunkInternal(LocalChunkProvider.java:443)
	at org.terasology.world.chunks.localChunkProvider.LocalChunkProvider.dispose(LocalChunkProvider.java:572)
	at org.terasology.engine.modes.StateIngame.dispose(StateIngame.java:117)
	at org.terasology.engine.TerasologyEngine.cleanup(TerasologyEngine.java:488)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:409)
	at org.terasology.engine.Terasology.main(Terasology.java:163)

```

--- a/engine/src/main/java/org/terasology/particles/components/ParticleEmitterComponent.java
+++ b/engine/src/main/java/org/terasology/particles/components/ParticleEmitterComponent.java
@@ -94,7 +94,7 @@ public class ParticleEmitterComponent implements Component {
     /**
      * This emitter's particle pool.
      */
-    public ParticlePool particlePool;
+    public transient ParticlePool particlePool;
 
     /**
      * Maps Generator component -> Function that processes that Generator

--- a/engine/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java
+++ b/engine/src/main/java/org/terasology/persistence/typeHandling/coreTypes/ObjectFieldMapTypeHandler.java
@@ -68,9 +68,14 @@ public class ObjectFieldMapTypeHandler<T> extends TypeHandler<T> {
 
             if (!Objects.equals(val, Defaults.defaultValue(field.getType()))) {
                 TypeHandler handler = entry.getValue();
-                PersistedData fieldValue = handler.serialize(val, serializer);
-                if (fieldValue != null) {
-                    mappedData.put(getFieldName(field), fieldValue);
+                try {
+                    PersistedData fieldValue = handler.serialize(val, serializer);
+                    if (fieldValue != null) {
+                        mappedData.put(getFieldName(field), fieldValue);
+                    }
+                } catch (StackOverflowError e) {
+                    logger.error("Likely circular reference in field {}.", field);
+                    throw e;
                 }
             }
         }

